src/controller/CHIPDevice.cpp

```diff
@@ -470,5 +496,29 @@ void Device::AddReportHandler(EndpointId endpoint, ClusterId cluster, AttributeI
     mCallbacksMgr.AddReportCallback(mDeviceId, endpoint, cluster, attribute, onReportCallback);
 }
 
+Device::~Device()
+{
+    if (mExchangeMgr)
+    {
+        // Ensure that any exchange contexts we have open get closed now,
+        // because we don't want them to call back in to us after this
+        // point.
+        mExchangeMgr->CloseAllContextsForDelegate(this);
+    }
+
+    if (mStorageDelegate)
+    {
+        // Store the current device in persistent storage so we have the latest
+        // message counters available next time.
+        Transport::PeerConnectionState * connectionState = mSessionManager->GetPeerConnectionState(mSecureSession);
+        VerifyOrReturn(connectionState != nullptr);
+        SerializedDevice serialized;
+        Serialize(serialized);
+        // TODO: no need to base-64 the serialized values AGAIN
```

Comment:
I thought our persistent data store interfaces were opaque binary now.  Why the need to base-64 encode at all?

Reply:
This TODO (which came with the copy-paste) is just talking about how the storage delegate ends up base64 encoding everything that gets passed in. In this case the device is already serialized to base64 and when we send it off to storage, it gets encoded a second time.

Reply:
Ah, I see.

So that's platform-specific behavior though, isn't it?  If so, I don't think this comment actually makes sense.

It's not a big deal though.